### PR TITLE
spec: allow re-declarations of identifiers and remove function result variables

### PR DESCRIPTION
**Pull request by Samyak2**

### CODE REVIEW

File: doc/bakugo_spec.md [hunk 1]
```diff
@@ -534,8 +534,9 @@ Blocks nest and influence _scoping_.
 | Changes yet to made |
```
Comment (ishwar00) [inline doc/bakugo_spec.md:534]: We can remove this _note_ now.

Comment (ishwar00) [inline doc/bakugo_spec.md:534]: We can remove this _note_ now.

File: doc/bakugo_spec.md [hunk 2]
```diff
@@ -546,15 +547,15 @@ The _scope_ of a declared identifier is the extent of source text in which the i
...
 
 
-Go is lexically scoped using _blocks_:
+Go is lexically scoped:
 
 1. The scope of a _predeclared identifier_ is the universe block.
-2. The scope of an identifier denoting a function parameter, or result variable is the function body.
-3. The scope of a constant or variable identifier declared inside a function begins at the end of the ConstSpec or VarSpec and ends at the end of the innermost containing block.
-4. The scope of a type identifier declared inside a function begins at the identifier in the TypeSpec and ends at the end of the innermost containing block.
+1. The scope of an identifier denoting a function result variable is the function body.
```
Comment (ishwar00) [inline doc/bakugo_spec.md:553]: Does this mean that an identifier denoting a function result variable cannot be re-declared in the function body?

File: doc/bakugo_spec.md (hunk 1 above)
Comment (Samyak2) [reply doc/bakugo_spec.md:534]: Ah yes. I didn't notice that

File: doc/bakugo_spec.md [hunk 3]
```diff
@@ -546,15 +544,15 @@ The _scope_ of a declared identifier is the extent of source text in which the i
...
```
Comment (ishwar00) [inline doc/bakugo_spec.md:550]: Does this mean that an identifier denoting a function result variable cannot be re-declared in the function body?

Comment (Samyak2) [reply doc/bakugo_spec.md:550]: Yes, that is correct. A function result variable is declared in the function definition and whatever assigned to it is returned. So it doesn't make sense to re-declare it anywhere in the body because the binding created at the definition is used at the end of the body.

Comment (Samyak2) [reply doc/bakugo_spec.md:550]: I never liked the idea of the return variable being declared in the definition. I don't mind removing that feature entirely xD

Comment (ishwar00) [reply doc/bakugo_spec.md:550]: > So it doesn't make sense to re-declare it anywhere in the body because the binding created at the definition is used at the end of the body.

Yes, I was wondering that. Cool! Implementing it would be fun.

Comment (ishwar00) [reply doc/bakugo_spec.md:550]: > I never liked the idea of the return variable being declared in the definition. I don't mind removing that feature entirely xD

Yes, I had something similar on my mind. xD

Comment (Samyak2) [reply doc/bakugo_spec.md:550]: Let's remove it then?

Comment (ishwar00) [reply doc/bakugo_spec.md]: Okay, Let's remove them.

Comment (Samyak2) [reply doc/bakugo_spec.md]: It's done in 4bce513

Comment (ishwar00) [reply doc/bakugo_spec.md]: Nice

File: doc/bakugo_spec.md [hunk 4]
```diff
@@ -414,22 +414,22 @@ rune        alias for int32
 
 ```
 Signature      = Parameters [ Result ] .
-Result         = Parameters | Type .
+Result         = Type .
```
Comment (ishwar00) [inline doc/bakugo_spec.md:417]:
```suggestion
Result         = "(" [ ResultList [ "," ] ] ")" | Type .
ResultList     = Type { "," Type } .
```
If I am not wrong, production `Result = Type .` will not allow the declaration of multiple return values.

The production name `ResultList` does not make much sense. I could not come up with a better one though.

File: doc/bakugo_spec.md [hunk 5]
```diff
@@ -1110,7 +1105,7 @@ if x := f(); x < y {
 ```
 
 ## Return statements
-A "return" statement in a function F terminates the execution of F, and optionally provides one or more result values.
+A "return" statement in a function F terminates the execution of F.
```
Comment (ishwar00) [inline doc/bakugo_spec.md:1108]: I noticed definition does not mention returning any values. Could you please explain why this change was made?

File: doc/bakugo_spec.md [hunk 6]
```diff
@@ -529,13 +529,11 @@ In addition to explicit blocks in the source code, there are implicit blocks:
...
 
 # Declarations and scope
-| NOTE |
-| :--- |
-| Changes yet to made |
 
 A _declaration_ binds a non-_blank_ identifier to a _constant_, _type_, _variable_, _function_. Every
-identifier in a program must be declared. No identifier may be declared twice in the same block, and no
-identifier may be declared in both the file and package block.
+identifier in a program must be declared. No identifier may be declared in both the file and package block.
```
Comment (ishwar00) [inline doc/bakugo_spec.md:534]:
```suggestion
identifier in a program must be declared.
```
Since we don't support _packages_ in this spec. I think we can remove it safely.

**Comment by ishwar00** [comment]
The current spec does not mention anything specific about the identifier declared at the top level(outside any function). Should we add it?

File: doc/bakugo_spec.md (hunk 4 above)
Comment (Samyak2) [reply doc/bakugo_spec.md:417]: Oh you're right.

> The production name ResultList does not make much sense. I could not come up with a better one though.

How about not making it a separate non-terminal?

```suggestion
Result         = "(" [ Type { "," Type } [ "," ] ] ")" | Type .
```

File: doc/bakugo_spec.md (hunk 5 above)
Comment (Samyak2) [reply doc/bakugo_spec.md:1108]: Oh good catch. I'll revert this

**Comment by Samyak2** [comment]
> The current spec does not mention anything specific about the identifier declared at the top level(outside any function). Should we add it?

Ah yes, we should. Could you point to it in the original spec if you remember it?

File: doc/bakugo_spec.md (hunk 4 above)
Comment (ishwar00) [reply doc/bakugo_spec.md:417]: Okay. It feels a bit crowded, but that will do.

**Comment by ishwar00** [comment]
Sure. We can refer
1. The scope of a _predeclared identifier_ is the universe block.
2. The scope of an identifier denoting a constant, type, variable, or function (but not method) declared at top level (outside any function) is the package block.

from original spec [Declarations and scope](https://go.dev/ref/spec#Declarations_and_scope). As we are not supporting packages in this spec, We can use universe block instead.

**Comment by ishwar00** [comment]
Hey @Samyak2,

I have added spec for top-level identifiers. Could you take a look?
